pennylane/ops/op_math/controlled_class.py

```diff
@@ -0,0 +1,380 @@
+
+        assert (
+            len(Wires.shared_wires([base.wires, control_wires])) == 0
+        ), "The control wires must be different from the base operation wires."
+
+        assert (
+            len(Wires.shared_wires([work_wires, base.wires + control_wires])) == 0
+        ), "Work wires must be different the control_wires and base operation wires."
+
+        self.hyperparameters["control_wires"] = control_wires
+        self.hyperparameters["control_values"] = control_values
+        self.hyperparameters["work_wires"] = work_wires
+
+        self._name = f"C{base.name}"
+
+        super().__init__(base, do_queue, id)
+
+    # Properties on the control values ######################
+    @property
+    def control_values(self):
+        """Iterable[Bool]. For each control wire, denotes whether to control on ``True`` or ``False``."""
+        return self.hyperparameters["control_values"]
+
+    @property
+    def _control_int(self):
+        """Int. Conversion of ``control_values`` to an integer."""
+        return sum(2**i for i, val in enumerate(reversed(self.control_values)) if val)
+
+    # Properties on the wires ##########################
+
+    @property
+    def control_wires(self):
+        """The control wires."""
+        return self.hyperparameters["control_wires"]
+
+    @property
+    def target_wires(self):
+        """The wires of the target operator."""
+        return self.base.wires
+
+    @property
+    def work_wires(self):
+        """Additional wires that can be used in the decomposition. Not modified by the operation."""
+        return self.hyperparameters["work_wires"]
+
+    @property
+    def wires(self):
+        return self.control_wires + self.base.wires + self.work_wires
+
+    # pylint: disable=protected-access
+    @property
+    def _wires(self):
+        return self.wires
+
+    # pylint: disable=protected-access
+    @_wires.setter
+    def _wires(self, new_wires):
+        new_wires = new_wires if isinstance(new_wires, Wires) else Wires(new_wires)
+
+        num_control = len(self.control_wires)
+        num_base = len(self.base.wires)
+        num_control_and_base = num_control + num_base
+
+        assert num_control_and_base <= len(new_wires), (
+            f"{self.name} needs at least {num_control_and_base} wires."
+            f" {len(new_wires)} provided."
+        )
+
+        self.hyperparameters["control_wires"] = new_wires[0:num_control]
+
+        self.base._wires = new_wires[num_control:num_control_and_base]
+
+        if len(new_wires) > num_control_and_base:
+            self.hyperparameters["work_wires"] = new_wires[num_control_and_base:]
+        else:
+            self.hyperparameters["work_wires"] = Wires([])
+
+    # Methods ##########################################
+
+    def label(self, decimals=None, base_label=None, cache=None):
+        return self.base.label(decimals=decimals, base_label=base_label, cache=cache)
```

Comment:
Shouldn't we add a `C` in the label? Meaning:
```python
import pennylane as qml
qml.ops.Controlled(qml.PauliX(0), 1).label()
```
should return `CX` and not `X`.